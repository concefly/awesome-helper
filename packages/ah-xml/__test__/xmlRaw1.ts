import * as fs from 'fs';

export const xmlRaw1 = fs.readFileSync('__test__/xml1.xml', { encoding: 'utf-8' });
