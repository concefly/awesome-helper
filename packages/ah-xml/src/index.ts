export * from './XmlHelper';
export * from './XmlNode';
export * from './QuerySelector';
