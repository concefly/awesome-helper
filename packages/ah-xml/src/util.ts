import { xml2js, Element } from 'xml-js';

export interface ElementWithId extends Element {
  id: string;
  parentId: string;
}

export interface IWalkXmlCtx {
  indexTrace: number[];
  parent?: Element;
}

export function walkXml(
  current: Element,
  tap: (c: Element, _ctx: IWalkXmlCtx) => void,
  ctx: IWalkXmlCtx = { indexTrace: [0] }
) {
  tap(current, ctx);

  current.elements?.forEach((child, index) => {
    walkXml(child, tap, {
      parent: current,
      indexTrace: ctx.indexTrace.concat(index),
    });
  });
}

export function convertXml2List(input: Element | string) {
  // 复制一遍
  const xml =
    typeof input === 'string'
      ? (xml2js(input, { compact: false }) as Element)
      : (JSON.parse(JSON.stringify(input)) as Element);

  const list: ElementWithId[] = [];

  walkXml(xml, (current, { parent, indexTrace }) => {
    const id = indexTrace.join('.');

    // walk 是深度遍历，parent 一定已经被打过 id 了
    const parentId = parent ? (parent as ElementWithId)!.id : '';

    Object.assign(current, { id, parentId });
    list.push(current as ElementWithId);
  });

  return list;
}

export function matchTrim<T extends string>(text: string, reg: RegExp, flag: T) {
  const match = text.match(reg);
  if (!match) return;

  text = text.replace(reg, '');

  return { text, match, flag };
}
