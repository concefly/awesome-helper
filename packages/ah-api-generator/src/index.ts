import path from 'path';
import fs from 'fs';
import { curl } from 'urllib';
import { renderService } from './util';
import ejs from 'ejs';
import { OpenAPIV3 as API } from 'openapi-types';
import prettier from 'prettier';
import RefParser from '@apidevtools/json-schema-ref-parser';

export interface IConfig {
  input?: { type: 'local'; filename: string } | { type: 'remote'; url: string; headers?: any };
  operationIdSplitter?: string;
  template?: string;
  banner?: string;
  dump?: string;
}

export const loadApiDoc = async (input: IConfig['input']): Promise<API.Document> => {
  if (!input) throw new Error('input 不存');

  let result =
    input.type === 'local'
      ? JSON.parse(fs.readFileSync(input.filename, { encoding: 'utf-8' }))
      : input.type === 'remote'
      ? await curl<any>(input.url, { headers: input.headers, dataType: 'json' }).then(r => r.data)
      : null;

  if (!result) throw new Error(`inputSource 不存在: ${JSON.stringify(input)}`);

  // 解 $ref
  result = await RefParser.dereference(result);

  return result;
};

export const generateAPIService = async (cfg: IConfig) => {
  const {
    input = { type: 'local', filename: 'api.json' },
    template = path.resolve(__dirname, '..', 'template.ejs'),
    dump,
    banner,
  } = cfg;

  const apiDoc = await loadApiDoc(input);
  const serviceMap = renderService(apiDoc, { operationIdSplitter: cfg.operationIdSplitter });

  // render template
  let content = ejs.render(fs.readFileSync(template, { encoding: 'utf-8' }), { serviceMap });
  if (banner) content = banner + '\n' + content;

  // pretty
  const prettierOpt = await prettier.resolveConfig(process.cwd());
  content = prettier.format(content, {
    parser: 'typescript',
    trailingComma: 'es5',
    tabWidth: 2,
    semi: true,
    singleQuote: true,
    jsxSingleQuote: true,
    printWidth: 100,
    bracketSpacing: true,
    arrowParens: 'avoid',
    ...prettierOpt,
  });

  // dump
  if (dump) fs.writeFileSync(dump, content, { encoding: 'utf-8' });

  return content;
};
