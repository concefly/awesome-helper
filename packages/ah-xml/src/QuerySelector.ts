import { matchTrim } from './util';
import { XmlNode, XmlElementNode } from './XmlNode';

/** 选择器 */
abstract class BaseSelector {
  abstract test(n: XmlNode): boolean;
}

/** 分组选择器 */
abstract class BaseGroupingSelector extends BaseSelector {}

/** 组合器 */
abstract class BaseCombinator {
  abstract filter(n: XmlNode): XmlNode[];
}

export type IQuerySelectorToken = BaseSelector | BaseGroupingSelector | BaseCombinator;

export class QuerySelector {
  constructor(readonly tokens: IQuerySelectorToken[]) {}

  addToken(t: IQuerySelectorToken) {
    const lastToken = this.tokens[this.tokens.length - 1];

    if (lastToken instanceof BaseCombinator) {
      if (t instanceof BaseCombinator) throw new Error('非法：组合器+组合器');
    }

    this.tokens.push(t);
    return this;
  }

  query(node: XmlNode) {
    let nodeList: XmlNode[] = [node].concat(node.allChildren);

    for (const t of this.tokens) {
      // BaseSelector or BaseGroupingSelector
      if (t instanceof BaseSelector || t instanceof BaseGroupingSelector) {
        nodeList = nodeList.filter(n => t.test(n));

        if (nodeList.length === 0) return [];
      }

      // BaseCombinator
      else if (t instanceof BaseCombinator) {
        let addNodeList: XmlNode[] = [];

        nodeList.forEach(n => {
          addNodeList = addNodeList.concat(t.filter(n));
        });

        nodeList = nodeList.concat(addNodeList);
      }
    }

    return nodeList;
  }
}

/** 元素选择器 */
export class TypeSelector extends BaseSelector {
  constructor(readonly tag: string) {
    super();
  }

  test(n: XmlNode) {
    if (n instanceof XmlElementNode) {
      return n.tagName.toUpperCase() === this.tag.toUpperCase();
    }

    return false;
  }
}

/** 属性选择器 */
export class AttributeSelector extends BaseSelector {
  constructor(readonly attrName: string, readonly valueReg?: RegExp) {
    super();
  }

  test(n: XmlNode) {
    if (!n.attributes) return false;
    if (typeof n.attributes.get(this.attrName) === 'undefined') return false;

    const value = n.attributes.get(this.attrName) + '';
    if (this.valueReg) return this.valueReg.test(value);

    return true;
  }
}

/** 选择器列表 */
export class SelectorList extends BaseGroupingSelector {
  constructor(readonly list: BaseSelector[]) {
    super();
  }

  test(n: XmlNode) {
    return this.list.some(s => s.test(n));
  }
}

/** 后代组合器 */
export class DescendantCombinator extends BaseCombinator {
  filter(n: XmlNode) {
    return n.allChildren;
  }
}

/** 直接子代组合器 */
export class ChildCombinator extends BaseCombinator {
  filter(n: XmlNode) {
    return n.flatChildren;
  }
}

export function createQuerySelector(selector: string): QuerySelector {
  // 规整 selector
  selector = selector.trim();
  selector = selector.replace(/\s+/g, ' ');

  const qs = new QuerySelector([]);

  let cnt = 99999;
  while (cnt-- > 0) {
    const mr =
      // selector
      matchTrim(selector, /^([\w_-]+)/, 'tag') ||
      matchTrim(selector, /^\[([\w_-]+)\]/, 'attr1') ||
      matchTrim(selector, /^\[([\w_-]+)=(.+?)\]/, 'attr2') ||
      // combinator
      matchTrim(selector, /^\s+(?![>])/, 'DescendantCombinator') ||
      matchTrim(selector, /^\s?\>\s?/, 'ChildCombinator');

    if (!mr) break;

    // 设置剩余部分
    selector = mr.text;

    switch (mr.flag) {
      case 'tag': {
        const tag = mr.match[1];
        qs.addToken(new TypeSelector(tag));
        break;
      }

      case 'attr1': {
        const attrName = mr.match[1];
        qs.addToken(new AttributeSelector(attrName));
        break;
      }

      case 'attr2': {
        const attrName = mr.match[1];
        const attrValue = mr.match[2];
        qs.addToken(new AttributeSelector(attrName, new RegExp(`^${attrValue}$`)));
        break;
      }

      case 'DescendantCombinator': {
        qs.addToken(new DescendantCombinator());
        break;
      }

      case 'ChildCombinator': {
        qs.addToken(new ChildCombinator());
        break;
      }

      default:
        throw new Error('不支持' + (mr as any).flag);
    }
  }

  if (selector.length) throw new Error(`"${selector}" 解析出错`);

  return qs;
}
