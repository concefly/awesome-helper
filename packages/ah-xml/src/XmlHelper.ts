import { TreeHelper } from 'ah-tree-helper';
import { convertXml2List } from './util';
import { XmlElementNode, XmlNode, XmlTextNode, XmlCDataNode } from './XmlNode';

function convertInput2NodeList(input: string) {
  const eleTypeMapper = new Map<string, typeof XmlNode>([
    ['text', XmlTextNode],
    ['element', XmlElementNode],
    ['cdata', XmlCDataNode],
  ]);

  return convertXml2List(input).map(ele => {
    const EleType = eleTypeMapper.get(ele.type!) || XmlNode;
    return new EleType(ele.id, ele.parentId, { xmlElement: ele });
  });
}

export class XmlHelper extends TreeHelper<XmlNode> {
  constructor(readonly raw: string) {
    super(convertInput2NodeList(raw));

    // 设置 tree
    this.list.forEach(t => t.setTree(this));
  }
}
