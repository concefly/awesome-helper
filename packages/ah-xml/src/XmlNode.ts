import { ITreeNode } from 'ah-tree-helper';
import { Element } from 'xml-js';
import { createQuerySelector } from './QuerySelector';
import { XmlHelper } from './XmlHelper';

type RestParameters<T extends (...args: any) => any> = T extends (_a0: any, ...args: infer P) => any
  ? P
  : never;

export class XmlNode implements ITreeNode {
  // 由 XmlHelper 初始化
  private tree!: XmlHelper;

  constructor(
    readonly id: string,
    readonly parentId: string,
    public meta: {
      xmlElement: Element;
    }
  ) {}

  setTree(tree: XmlHelper) {
    this.tree = tree;
    return this;
  }

  get flatChildren() {
    return this.tree.getFlatChildren(this.id);
  }

  get allChildren() {
    const list: XmlNode[] = [];
    this.walk(c => {
      if (c.id === this.id) return;
      list.push(c);
    });

    return list;
  }

  get parent() {
    return this.tree.getById(this.parentId);
  }

  get attributes() {
    return new Map<string, string>(
      Object.entries(this.meta.xmlElement.attributes || {}).map(([n, v]) => [n, v + ''])
    );
  }

  walk(...args: RestParameters<XmlHelper['walk']>) {
    return this.tree.walk(this.id, ...args);
  }

  query(selector: string) {
    const qs = createQuerySelector(selector);
    return qs.query(this);
  }

  toString() {
    return `<${this.meta.xmlElement.name || '_UnknownTag'} />`;
  }
}

export class XmlElementNode extends XmlNode {
  get tagName() {
    return this.meta.xmlElement.name!;
  }

  toString() {
    const attrStr = [['__id', this.id], ...Array.from(this.attributes.entries())]
      .map(([n, v]) => `${n}="${v}"`)
      .join(' ');

    if (this.flatChildren.length === 0) {
      return `<${this.tagName} ${attrStr}/>`;
    }

    return `<${this.tagName} ${attrStr}>...<${this.tagName}/>`;
  }
}

export class XmlTextNode extends XmlNode {
  get tagName() {
    return '';
  }

  get text() {
    return this.meta.xmlElement.text + '';
  }

  toString() {
    return this.meta.xmlElement.text + '';
  }
}

export class XmlCDataNode extends XmlNode {
  get tagName() {
    return '';
  }

  get cdata() {
    return this.meta.xmlElement.cdata || '';
  }

  toString() {
    return 'cdata';
  }
}
