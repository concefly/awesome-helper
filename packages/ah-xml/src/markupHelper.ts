import { XmlElementNode, XmlNode, XmlTextNode } from './XmlNode';

export function noteToHTML(
  node: XmlNode,
  opt?: {
    tapAttrs?(attrs: [string, string][], current: XmlNode): [string, string][];
  }
): string {
  const selfCloseTagSet = new Set<string>(['br', 'img']);

  const convertXml2Html = (current: XmlNode): string => {
    if (current instanceof XmlElementNode) {
      const tag = current.tagName;

      let attributes = Array.from(current.attributes.entries());

      if (opt?.tapAttrs) {
        attributes = opt.tapAttrs(attributes, current);
      }

      const attrStr = attributes.map(([_n, _v]) => `${_n}="${_.escape(_v)}"`).join(' ');
      const isSelfClose = selfCloseTagSet.has(tag) && current.flatChildren.length === 0;

      if (isSelfClose) {
        return `<${tag} ${attrStr} />`;
      }

      const childrenStr = current.flatChildren.map(convertXml2Html).join('\n');
      return `<${tag} ${attrStr}>${childrenStr}</${tag}>`;
    }

    if (current instanceof XmlTextNode) return current.text;

    return '';
  };

  return convertXml2Html(node);
}
