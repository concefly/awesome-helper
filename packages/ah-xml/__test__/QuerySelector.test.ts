import { createQuerySelector, XmlHelper } from '../src';
import { xmlRaw1 } from './xmlRaw1';

const xml = new XmlHelper(xmlRaw1);
const xmlRoot = xml.findAllRoot()[0].flatChildren[1];

describe('createQuerySelector', () => {
  [
    'div',
    'div   input',
    'div > input',
    'div[title=aa]',
    'div[title=aa] input',
    'div[title=aa] input[type]',
    'div[title=aa] input[type]    >  span',
  ].forEach(selector => {
    const humanizeSelector = `"${selector}"`;

    it(humanizeSelector, () => {
      const qs = createQuerySelector(selector);
      expect(qs.tokens).toMatchSnapshot();
    });
  });
});

describe('QuerySelector', () => {
  [
    'breakfast_menu food name',
    'breakfast_menu food[title=菜名]',
    'breakfast_menu food[title]',
    'breakfast_menu food',
    'food > name[d=1]',
    'name[d]',
    '[title] [d]',
    '[title] [d=1]',
    '[title=菜名] [d=2]',
  ].forEach(selector => {
    const humanizeSelector = `"${selector}"`;

    it(humanizeSelector, () => {
      const res = xmlRoot.query(selector);

      expect(res.map(r => r.toString())).toMatchSnapshot();
    });
  });
});
