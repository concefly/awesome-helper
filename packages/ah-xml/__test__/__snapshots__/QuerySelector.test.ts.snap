// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuerySelector "[title=菜名] [d=2]" 1`] = `Array []`;

exports[`QuerySelector "[title] [d=1]" 1`] = `
Array [
  "<name __id=\\"0.1.0.0\\" d=\\"1\\">...<name/>",
  "<price __id=\\"0.1.0.1\\" d=\\"1\\">...<price/>",
  "<description __id=\\"0.1.0.2\\" d=\\"1\\">...<description/>",
  "<calories __id=\\"0.1.0.3\\" d=\\"1\\">...<calories/>",
  "<name __id=\\"0.1.0.0\\" d=\\"1\\">...<name/>",
  "<price __id=\\"0.1.0.1\\" d=\\"1\\">...<price/>",
  "<description __id=\\"0.1.0.2\\" d=\\"1\\">...<description/>",
  "<calories __id=\\"0.1.0.3\\" d=\\"1\\">...<calories/>",
]
`;

exports[`QuerySelector "[title] [d]" 1`] = `
Array [
  "<name __id=\\"0.1.0.0\\" d=\\"1\\">...<name/>",
  "<price __id=\\"0.1.0.1\\" d=\\"1\\">...<price/>",
  "<description __id=\\"0.1.0.2\\" d=\\"1\\">...<description/>",
  "<calories __id=\\"0.1.0.3\\" d=\\"1\\">...<calories/>",
  "<name __id=\\"0.1.1.0\\" d=\\"2\\">...<name/>",
  "<price __id=\\"0.1.1.1\\" d=\\"2\\">...<price/>",
  "<description __id=\\"0.1.1.2\\" d=\\"2\\">...<description/>",
  "<calories __id=\\"0.1.1.3\\" d=\\"2\\">...<calories/>",
  "<name __id=\\"0.1.0.0\\" d=\\"1\\">...<name/>",
  "<price __id=\\"0.1.0.1\\" d=\\"1\\">...<price/>",
  "<description __id=\\"0.1.0.2\\" d=\\"1\\">...<description/>",
  "<calories __id=\\"0.1.0.3\\" d=\\"1\\">...<calories/>",
]
`;

exports[`QuerySelector "breakfast_menu food name" 1`] = `
Array [
  "<name __id=\\"0.1.0.0\\" d=\\"1\\">...<name/>",
  "<name __id=\\"0.1.1.0\\" d=\\"2\\">...<name/>",
]
`;

exports[`QuerySelector "breakfast_menu food" 1`] = `
Array [
  "<food __id=\\"0.1.0\\" title=\\"菜名\\">...<food/>",
  "<food __id=\\"0.1.1\\">...<food/>",
]
`;

exports[`QuerySelector "breakfast_menu food[title=菜名]" 1`] = `
Array [
  "<food __id=\\"0.1.0\\" title=\\"菜名\\">...<food/>",
]
`;

exports[`QuerySelector "breakfast_menu food[title]" 1`] = `
Array [
  "<food __id=\\"0.1.0\\" title=\\"菜名\\">...<food/>",
]
`;

exports[`QuerySelector "food > name[d=1]" 1`] = `
Array [
  "<name __id=\\"0.1.0.0\\" d=\\"1\\">...<name/>",
]
`;

exports[`QuerySelector "name[d]" 1`] = `
Array [
  "<name __id=\\"0.1.0.0\\" d=\\"1\\">...<name/>",
  "<name __id=\\"0.1.1.0\\" d=\\"2\\">...<name/>",
]
`;

exports[`createQuerySelector "div   input" 1`] = `
Array [
  TypeSelector {
    "tag": "div",
  },
  DescendantCombinator {},
  TypeSelector {
    "tag": "input",
  },
]
`;

exports[`createQuerySelector "div > input" 1`] = `
Array [
  TypeSelector {
    "tag": "div",
  },
  ChildCombinator {},
  TypeSelector {
    "tag": "input",
  },
]
`;

exports[`createQuerySelector "div" 1`] = `
Array [
  TypeSelector {
    "tag": "div",
  },
]
`;

exports[`createQuerySelector "div[title=aa] input" 1`] = `
Array [
  TypeSelector {
    "tag": "div",
  },
  AttributeSelector {
    "attrName": "title",
    "valueReg": /\\^aa\\$/,
  },
  DescendantCombinator {},
  TypeSelector {
    "tag": "input",
  },
]
`;

exports[`createQuerySelector "div[title=aa] input[type]    >  span" 1`] = `
Array [
  TypeSelector {
    "tag": "div",
  },
  AttributeSelector {
    "attrName": "title",
    "valueReg": /\\^aa\\$/,
  },
  DescendantCombinator {},
  TypeSelector {
    "tag": "input",
  },
  AttributeSelector {
    "attrName": "type",
    "valueReg": undefined,
  },
  ChildCombinator {},
  TypeSelector {
    "tag": "span",
  },
]
`;

exports[`createQuerySelector "div[title=aa] input[type]" 1`] = `
Array [
  TypeSelector {
    "tag": "div",
  },
  AttributeSelector {
    "attrName": "title",
    "valueReg": /\\^aa\\$/,
  },
  DescendantCombinator {},
  TypeSelector {
    "tag": "input",
  },
  AttributeSelector {
    "attrName": "type",
    "valueReg": undefined,
  },
]
`;

exports[`createQuerySelector "div[title=aa]" 1`] = `
Array [
  TypeSelector {
    "tag": "div",
  },
  AttributeSelector {
    "attrName": "title",
    "valueReg": /\\^aa\\$/,
  },
]
`;
